### PR TITLE
[multi head attention] Added multi head attention scaffold @open sesame 8/22 13:32

### PR DESCRIPTION
**Pull request by lhs8928**
- Added calcCommonDerivative
     - Added finalizeCommon
     - Replace torch_ref property to reference_framework
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

### CODE REVIEW

File: nntrainer/layers/multi_head_attention_layer.h [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+#include <acti_func.h>
+#include <layer_impl.h>
+
+namespace nntrainer {
+
+/**
+ * @class   Multi Head Attention Layer
+ * @brief   Implementation of multi head attention which is described in paper
+ * "Attention is all you need"
```
Comment (jihochu) [inline nntrainer/layers/multi_head_attention_layer.h:27]: How about adding url for paper? https://arxiv.org/abs/1706.03762

File: nntrainer/layers/multi_head_attention_layer.h [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+public:
+  /**
+   * @brief     Constructor of MultiHeadAttention Layer
+   */
+  MultiHeadAttentionLayer();
+
+  /**
+   * @brief     Destructor of MultiHeadAttention Layer
+   */
+  ~MultiHeadAttentionLayer();
```
Comment (jihochu) [inline nntrainer/layers/multi_head_attention_layer.h:39]: If MultiHeadAttentionLayer is not final layer, it's better to make destructor to virtual method.

File: nntrainer/layers/multi_head_attention_layer.h (hunk 1 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.h:27]: Already described at line 8

File: nntrainer/layers/multi_head_attention_layer.h (hunk 2 above)
Comment (lhs8928) [reply nntrainer/layers/multi_head_attention_layer.h:39]: I guess MultiHeadAttentionLayer would be the final layer

File: nntrainer/layers/common_properties.h [hunk 3]
```diff
@@ -80,6 +80,27 @@ class Name : public nntrainer::Property<std::string> {
   bool isValid(const std::string &v) const override;
 };
 
+/**
+ * @brief     Enumeration of reference framework
+ */
+struct ReferenceFrameworkInfo {
+  enum class Enum { pytorch, tensorflow };
```
Comment (jijoongmoon) [inline nntrainer/layers/common_properties.h:87]: I think `torch_ref` property of adam optimizer needs to be fixed to be compatible with it. Also need to check if it is ok to change the this signature in API as well.

File: nntrainer/layers/common_properties.h [hunk 4]
```diff
@@ -1115,6 +1136,138 @@ class MoL_K : public PositiveIntegerProperty {
...
+  static constexpr const char *key = "value_dim"; /**< unique key to access */
+  using prop_tag = uint_prop_tag;                 /**< property type */
+};
+
+/**
+ * @brief ProjectedKeyDim property, projected key dim in multi head attention
+ * @details Correspond with key_dim of tensorflow
+ *
+ */
+class ProjectedKeyDim : public PositiveIntegerProperty {
```
Comment (jijoongmoon) [inline nntrainer/layers/common_properties.h:1192]: I wonder what is the difference betwwen `KeyDim` and `ProjectedKeyDim`?

Comment (lhs8928) [reply nntrainer/layers/common_properties.h:1192]: It is not a alias and used for actual different purpose.
In multi head attention layer each input(query, key, value) will be projected by 3 different fc layer.(query_fc, key_fc, value_fc)
The KeyDim means input key width dimension which also means input size of key_fc layer.
The projectedKeyDim means projected key width dimension which also means unit size of key_fc layer.

There is 2 different algorithm in multi head attention layer of torch/tensorflow.
1. If the return_attention_weight is set tensorflow return attention weight before applying dropout whereas pytorch will return the attention weight which is already applied the dropout.
2. The true/false value behaves differently in attention mask. In tensorflow 1 indicates attention and 0 indicates no attention whereas 1 indicates no attention and 0 indicates attention in pytorch

File: nntrainer/layers/common_properties.h (hunk 3 above)
Comment (lhs8928) [reply nntrainer/layers/common_properties.h:87]: In this commit torch_ref property of adam optimizer already compatible with ReferenceFrameworkInfo.
Let me check is it okay to change the API.

File: nntrainer/layers/common_properties.h (hunk 4 above)
Comment (jijoongmoon) [reply nntrainer/layers/common_properties.h:1194]: Like we discussted off-line, nntrainer does not need to work as other frameworks. The compatibility with other frameworks can be considered in the model loader and if it is possible to convert them in model loader level, then we don't need to implement like this. Optimizer Adam is different. It is the case which is different the algorithm itself.

Comment (lhs8928) [reply nntrainer/layers/common_properties.h:1141]: Done

File: nntrainer/layers/common_properties.cpp [hunk 5]
```diff
@@ -333,6 +333,17 @@ void GenericShape::set(const TensorDim &value) {
...
 }
 
+NumHeads::NumHeads(unsigned int value) { set(value); }
+
+ProvideAttentionMask::ProvideAttentionMask(bool value) { set(value); }
+
+ReturnAttentionWeight::ReturnAttentionWeight(bool value) { set(value); }
+
+ReturnAttentionWeightBeforeDropout::ReturnAttentionWeightBeforeDropout(
+  bool value) {
```
Comment (jijoongmoon) [inline nntrainer/layers/common_properties.cpp:343]: What if those two properties set at the same time?

Comment (lhs8928) [reply nntrainer/layers/common_properties.cpp]: Discussed offline to use only ReturnAttentionWeight property.
ReturnAttentionWeight will be string type property which will have one of "empty", "before", "after" value.

**Review by jijoongmoon — APPROVED** [review]
LGTM. One thing small thing is, we need to add more detail explanation about `ReturnAttentionWeight` Property. Such as, `"before"`  means `"return weight before mask is applied."`

**Comment by lhs8928** [comment]
> LGTM. One thing small thing is, we need to add more detail explanation about `ReturnAttentionWeight` Property. Such as, `"before"` means `"return weight before mask is applied."`

I describe it in https://github.com/nnstreamer/nntrainer/pull/1956/files#diff-f94da3514caf5dd909901e06dd23b5e75b237b9cf702d288862cd0918ccff7e7R1192

**Review by DonghakPark — APPROVED** [review]
LGTM

**Review by songgot — APPROVED** [review]
LGTM

**Comment by jijoongmoon** [comment]
> > LGTM. One thing small thing is, we need to add more detail explanation about `ReturnAttentionWeight` Property. Such as, `"before"` means `"return weight before mask is applied."`
> 
> I describe it in https://github.com/nnstreamer/nntrainer/pull/1956/files#diff-f94da3514caf5dd909901e06dd23b5e75b237b9cf702d288862cd0918ccff7e7R1192

What I mean is for the APIs for user's documentation.
